core/drivers/crypto/hisilicon/hpre_rsa.c

```diff
@@ -0,0 +1,1076 @@
+		EMSG("Recv task error, ret=%d", ret);
+		res = TEE_ERROR_BAD_STATE;
+		goto done_proc;
+	}
+
+done_proc:
+	hisi_qm_release_qp(rsa_qp);
+
+	return res;
+}
+
+static TEE_Result mgf_process(size_t digest_size, uint8_t *seed,
+			      size_t seed_len, uint8_t *mask, size_t mask_len,
+			      struct drvcrypt_rsa_ed *rsa_data)
+{
+	struct drvcrypt_rsa_mgf mgf = { };
+
+	if (!rsa_data->mgf) {
+		EMSG("mgf function is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	mgf.hash_algo = rsa_data->hash_algo;
+	mgf.digest_size = digest_size;
+	mgf.seed.data = seed;
+	mgf.seed.length = seed_len;
+	mgf.mask.data = mask;
+	mgf.mask.length = mask_len;
+
+	return rsa_data->mgf(&mgf);
+}
+
+static TEE_Result xor_process(uint8_t *a, uint8_t *b, uint8_t *out, size_t len)
+{
+	struct drvcrypt_mod_op xor_mod = { };
+
+	xor_mod.n.length = len;
+	xor_mod.a.data = a;
+	xor_mod.a.length = len;
+	xor_mod.b.data = b;
+	xor_mod.b.length = len;
+	xor_mod.result.data = out;
+	xor_mod.result.length = len;
+
+	return drvcrypt_xor_mod_n(&xor_mod);
+}
+
+static TEE_Result hpre_rsa_allocate_publickey(struct rsa_public_key *key,
+					      size_t size_bits)
+{
+	if (!key) {
+		EMSG("Invalid input parameter");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	memset(key, 0, sizeof(*key));
+	key->e = crypto_bignum_allocate(size_bits);
+	if (!key->e)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	key->n = crypto_bignum_allocate(size_bits);
+	if (!key->n) {
+		crypto_bignum_free(&key->e);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void hpre_rsa_free_publickey(struct rsa_public_key *key)
+{
+	if (key) {
+		crypto_bignum_free(&key->e);
+		crypto_bignum_free(&key->n);
+	}
+}
+
+static size_t hpre_rsa_get_hw_kbytes(size_t key_bits)
+{
+	size_t size = 0;
+
+	if (key_bits <= 1024)
+		size = BITS_TO_BYTES(1024);
+	else if (key_bits <= 2048)
+		size = BITS_TO_BYTES(2048);
+	else if (key_bits <= 3072)
+		size = BITS_TO_BYTES(3072);
+	else if (key_bits <= 4096)
+		size = BITS_TO_BYTES(4096);
+	else
+		EMSG("Invalid key_bits[%zu]", key_bits);
+
+	return size;
+}
+
+static void hpre_rsa_params_free(struct hpre_rsa_msg *msg)
+{
+	size_t len = 0;
+
+	switch (msg->alg_type) {
+	case HPRE_ALG_KG_CRT:
+		free_wipe(msg->in);
+		break;
+	case HPRE_ALG_NC_NCRT:
+		if (msg->is_private)
+			free_wipe(msg->prikey);
+		else
+			free(msg->pubkey);
+		break;
+	case HPRE_ALG_NC_CRT:
+		if (msg->is_private) {
+			len = HPRE_RSA_CRT_KEY_BUF_SIZE(msg->key_bytes);
+			free_wipe(msg->prikey);
+		}
+		break;
+	default:
+		EMSG("Invalid alg_type[%"PRIu32"]", msg->alg_type);
+		break;
+	}
+}
+
+static enum hisi_drv_status hpre_rsa_encrypt_alloc(struct hpre_rsa_msg *msg)
+{
+	uint32_t size = HPRE_RSA_NCRT_TOTAL_BUF_SIZE(msg->key_bytes);
+	uint8_t *data = NULL;
+
+	data = calloc(1, size);
+	if (!data) {
+		EMSG("Fail to alloc rsa ncrt buf");
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+
+	msg->pubkey = data;
+	msg->pubkey_dma = virt_to_phys(msg->pubkey);
+
+	msg->in = data + (msg->key_bytes << 1);
```

Comment:
Why do you use `<< 1` instead of `* 2`?
I prefer `* 2` since we care about the value here, not the position of the bits.

Reply:
solved